crates/typst/src/layout/grid/mod.rs

```diff
@@ -0,0 +1,370 @@
+        styles: StyleChain,
+        regions: Regions,
+    ) -> SourceResult<Fragment> {
+        let inset = self.inset(styles);
+        let align = self.align(styles);
+        let columns = self.columns(styles);
+        let rows = self.rows(styles);
+        let column_gutter = self.column_gutter(styles);
+        let row_gutter = self.row_gutter(styles);
+        let fill = self.fill(styles);
+        let stroke = self.stroke(styles).map(Stroke::unwrap_or_default);
+
+        let tracks = Axes::new(columns.0.as_slice(), rows.0.as_slice());
+        let gutter = Axes::new(column_gutter.0.as_slice(), row_gutter.0.as_slice());
+        let grid = CellGrid::resolve(
+            tracks,
+            gutter,
+            self.children().clone(),
+            fill,
+            align,
+            inset,
+            engine,
+            styles,
+        )?;
+
+        let layouter = GridLayouter::new(&grid, &stroke, regions, styles, self.span());
+
+        // Measure the columns and layout the grid row-by-row.
+        Ok(layouter.layout(engine)?.fragment)
+    }
+}
+
+/// Track sizing definitions.
+#[derive(Debug, Default, Clone, Eq, PartialEq, Hash)]
+pub struct TrackSizings(pub SmallVec<[Sizing; 4]>);
+
+cast! {
+    TrackSizings,
+    self => self.0.into_value(),
+    sizing: Sizing => Self(smallvec![sizing]),
+    count: NonZeroUsize => Self(smallvec![Sizing::Auto; count.get()]),
+    values: Array => Self(values.into_iter().map(Value::cast).collect::<StrResult<_>>()?),
+}
+
+/// A cell in the grid. Use this to either override grid properties for a
+/// particular cell, or in show rules to apply certain styles to multiple cells
+/// at once.
+///
+/// For example, you can override the fill, alignment or inset for a single
+/// cell:
+///
+/// ```example
+/// #grid(
+///   columns: 2,
+///   fill: red,
+///   align: left,
+///   inset: 5pt,
+///   [ABC], [ABC],
+///   grid.cell(fill: blue)[C], [D],
+///   grid.cell(align: center)[E], [F],
+///   [G], grid.cell(inset: 0pt)[H]
+/// )
+/// ```
+#[elem(name = "cell", title = "Grid Cell", Show)]
+pub struct GridCell {
+    /// The cell's body.
+    #[required]
+    body: Content,
+
+    /// The cell's fill override.
+    fill: Smart<Option<Paint>>,
+
+    /// The cell's alignment override.
+    align: Smart<Align>,
+
+    /// The cell's inset override.
+    inset: Smart<Sides<Option<Rel<Length>>>>,
+}
+
+cast! {
+    GridCell,
+    v: Content => v.into(),
+}
+
+impl ResolvableCell for GridCell {
+    fn resolve_cell(
+        mut self,
+        _: usize,
+        _: usize,
+        fill: &Option<Paint>,
+        align: Smart<Align>,
+        inset: Sides<Rel<Length>>,
+        styles: StyleChain,
+    ) -> Cell {
+        let fill = self.fill(styles).unwrap_or_else(|| fill.clone());
+        self.push_fill(Smart::Custom(fill.clone()));
+        self.push_align(self.align(styles).or(align));
+        self.push_inset(self.inset(styles).or_else(|| Smart::Custom(inset.map(Some))));
```

Comment:
Should we fold here? (E.g.: on a `grid(align: left, ...)`, a cell with `grid.cell(align: top)` would be aligned at `top + left` instead of `top`)

Reply:
I think that would make sense.